src/control/system/database.go

```diff
@@ -0,0 +1,673 @@
+//
+// (C) Copyright 2020 Intel Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package system
+
+import (
+	"context"
+	"net"
+	"os"
+	"path/filepath"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/hashicorp/raft"
+	raftboltdb "github.com/hashicorp/raft-boltdb"
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+const (
+	sysDBFile            = "daos_system.db"
+	CurrentSchemaVersion = 0
+)
+
+type (
+	onLeadershipGainedFn func(context.Context) error
+	onLeadershipLostFn   func() error
```

Comment:
The gained leadership callbacks accept a child context that is created in the leadership monitoring loop. When leadership is lost, the child context is canceled, giving us an easy way to clean up any goroutines that were spawned as a result of gaining leadership. The callbacks for losing leadership shouldn't be spawning goroutines or otherwise doing things that need to be canceled.

Reply:
Ah, that's fair.